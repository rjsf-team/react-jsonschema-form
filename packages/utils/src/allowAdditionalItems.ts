import isObject from './isObject';
import { RJSFSchema, StrictRJSFSchema } from './types';

/** Checks the schema to see if it is allowing additional items, by verifying that `schema.additionalItems` is an
 * object. The user is warned in the console if `schema.additionalItems` has the value `true`.
 *
 * @param schema - The schema object to check
 * @returns - True if additional items is allowed, otherwise false
 */
export default function allowAdditionalItems<S extends StrictRJSFSchema = RJSFSchema>(schema: S) {
  if (schema.additionalItems === true) {
    console.warn('additionalItems=true is currently not supported');
  }
  return isObject(schema.additionalItems);
}
