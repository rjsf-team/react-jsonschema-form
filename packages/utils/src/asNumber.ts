/** Attempts to convert the string into a number. If an empty string is provided, then `undefined` is returned. If a
 * `null` is provided, it is returned. If the string ends in a `.` then the string is returned because the user may be
 * in the middle of typing a float number. If a number ends in a pattern like `.0`, `.20`, `.030`, string is returned
 * because the user may be typing number that will end in a non-zero digit. Otherwise, the string is wrapped by
 * `Number()` and if that result is not `NaN`, that number will be returned, otherwise the string `value` will be.
 *
 * @param value - The string or null value to convert to a number
 * @returns - The `value` converted to a number when appropriate, otherwise the `value`
 */
export default function asNumber(value: string | null) {
  if (value === '') {
    return undefined;
  }
  if (value === null) {
    return null;
  }
  if (/\.$/.test(value)) {
    // '3.' can't really be considered a number even if it parses in js. The
    // user is most likely entering a float.
    return value;
  }
  if (/\.0$/.test(value)) {
    // we need to return this as a string here, to allow for input like 3.07
    return value;
  }

  if (/\.\d*0$/.test(value)) {
    // It's a number, that's cool - but we need it as a string so it doesn't screw
    // with the user when entering dollar amounts or other values (such as those with
    // specific precision or number of significant digits)
    return value;
  }

  const n = Number(value);
  const valid = typeof n === 'number' && !Number.isNaN(n);

  return valid ? n : value;
}
