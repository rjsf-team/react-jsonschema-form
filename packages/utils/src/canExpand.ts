import { FormContextType, RJSFSchema, StrictRJSFSchema, UiSchema } from './types';
import getUiOptions from './getUiOptions';

/** Checks whether the field described by `schema`, having the `uiSchema` and `formData` supports expanding. The UI for
 * the field can expand if it has additional properties, is not forced as non-expandable by the `uiSchema` and the
 * `formData` object doesn't already have `schema.maxProperties` elements.
 *
 * @param schema - The schema for the field that is being checked
 * @param [uiSchema={}] - The uiSchema for the field
 * @param [formData] - The formData for the field
 * @returns - True if the schema element has additionalProperties, is expandable, and not at the maxProperties limit
 */
export default function canExpand<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  schema: RJSFSchema,
  uiSchema: UiSchema<T, S, F> = {},
  formData?: T
) {
  if (!schema.additionalProperties) {
    return false;
  }
  const { expandable = true } = getUiOptions<T, S, F>(uiSchema);
  if (expandable === false) {
    return expandable;
  }
  // if ui:options.expandable was not explicitly set to false, we can add
  // another property if we have not exceeded maxProperties yet
  if (schema.maxProperties !== undefined && formData) {
    return Object.keys(formData).length < schema.maxProperties;
  }
  return true;
}
