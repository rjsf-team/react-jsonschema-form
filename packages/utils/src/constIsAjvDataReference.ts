import { CONST_KEY, getSchemaType, isObject } from './';
import { RJSFSchema, StrictRJSFSchema } from './types';
import { JSONSchema7Type } from 'json-schema';
import isString from 'lodash/isString';

/**
 * Checks if the schema const property value is an AJV $data reference
 * and the current schema is not an object or array
 *
 * @param schema - The schema to check if the const is an AJV $data reference
 * @returns - true if the schema const property value is an AJV $data reference otherwise false.
 */
export default function constIsAjvDataReference<S extends StrictRJSFSchema = RJSFSchema>(schema: S): boolean {
  const schemaConst = schema[CONST_KEY] as JSONSchema7Type & { $data: string };
  const schemaType = getSchemaType<S>(schema);
  return isObject(schemaConst) && isString(schemaConst?.$data) && schemaType !== 'object' && schemaType !== 'array';
}
