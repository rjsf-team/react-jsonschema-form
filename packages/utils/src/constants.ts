/** Below are the list of all the keys into various elements of a RJSFSchema or UiSchema that are used by the various
 * utility functions. In addition to those keys, there are the special `ADDITIONAL_PROPERTY_FLAG` and
 * `RJSF_ADDITONAL_PROPERTIES_FLAG` flags that is added to a schema under certain conditions by the `retrieveSchema()`
 * utility.
 */
export const ADDITIONAL_PROPERTY_FLAG = '__additional_property';
export const ADDITIONAL_PROPERTIES_KEY = 'additionalProperties';
export const ALL_OF_KEY = 'allOf';
export const ANY_OF_KEY = 'anyOf';
export const CONST_KEY = 'const';
export const DEFAULT_KEY = 'default';
export const DEFINITIONS_KEY = 'definitions';
export const DEPENDENCIES_KEY = 'dependencies';
export const ENUM_KEY = 'enum';
export const ERRORS_KEY = '__errors';
export const ID_KEY = '$id';
export const IF_KEY = 'if';
export const ITEMS_KEY = 'items';
export const JUNK_OPTION_ID = '_$junk_option_schema_id$_';
export const NAME_KEY = '$name';
export const ONE_OF_KEY = 'oneOf';
export const PROPERTIES_KEY = 'properties';
export const REQUIRED_KEY = 'required';
export const SUBMIT_BTN_OPTIONS_KEY = 'submitButtonOptions';
export const REF_KEY = '$ref';
export const RJSF_ADDITONAL_PROPERTIES_FLAG = '__rjsf_additionalProperties';
export const ROOT_SCHEMA_PREFIX = '__rjsf_rootSchema';
export const UI_FIELD_KEY = 'ui:field';
export const UI_WIDGET_KEY = 'ui:widget';
export const UI_OPTIONS_KEY = 'ui:options';
export const UI_GLOBAL_OPTIONS_KEY = 'ui:globalOptions';
