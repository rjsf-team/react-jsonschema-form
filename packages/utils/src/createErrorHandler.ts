import isPlainObject from 'lodash/isPlainObject';

import { ERRORS_KEY } from './constants';
import { FieldValidation, FormValidation, GenericObjectType } from './types';

/** Given a `formData` object, recursively creates a `FormValidation` error handling structure around it
 *
 * @param formData - The form data around which the error handler is created
 * @returns - A `FormValidation` object based on the `formData` structure
 */
export default function createErrorHandler<T = any>(formData: T): FormValidation<T> {
  const handler: FieldValidation = {
    // We store the list of errors for this node in a property named __errors
    // to avoid name collision with a possible sub schema field named
    // 'errors' (see `utils.toErrorSchema`).
    [ERRORS_KEY]: [],
    addError(message: string) {
      this[ERRORS_KEY]!.push(message);
    },
  };
  if (Array.isArray(formData)) {
    return formData.reduce((acc, value, key) => {
      return { ...acc, [key]: createErrorHandler(value) };
    }, handler);
  }
  if (isPlainObject(formData)) {
    const formObject: GenericObjectType = formData as GenericObjectType;
    return Object.keys(formObject).reduce((acc, key) => {
      return { ...acc, [key]: createErrorHandler(formObject[key]) };
    }, handler as FormValidation<T>);
  }
  return handler as FormValidation<T>;
}
