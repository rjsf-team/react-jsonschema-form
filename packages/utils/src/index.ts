import allowAdditionalItems from './allowAdditionalItems';
import asNumber from './asNumber';
import canExpand from './canExpand';
import createErrorHandler from './createErrorHandler';
import createSchemaUtils from './createSchemaUtils';
import dataURItoBlob from './dataURItoBlob';
import deepEquals from './deepEquals';
import englishStringTranslator from './englishStringTranslator';
import enumOptionsDeselectValue from './enumOptionsDeselectValue';
import enumOptionsIndexForValue from './enumOptionsIndexForValue';
import enumOptionsIsSelected from './enumOptionsIsSelected';
import enumOptionsSelectValue from './enumOptionsSelectValue';
import enumOptionsValueForIndex from './enumOptionsValueForIndex';
import ErrorSchemaBuilder from './ErrorSchemaBuilder';
import findSchemaDefinition from './findSchemaDefinition';
import getDiscriminatorFieldFromSchema from './getDiscriminatorFieldFromSchema';
import getInputProps from './getInputProps';
import getSchemaType from './getSchemaType';
import getSubmitButtonOptions from './getSubmitButtonOptions';
import getTemplate from './getTemplate';
import getUiOptions from './getUiOptions';
import getWidget from './getWidget';
import guessType from './guessType';
import hashForSchema from './hashForSchema';
import hasWidget from './hasWidget';
import { ariaDescribedByIds, descriptionId, errorId, examplesId, helpId, optionId, titleId } from './idGenerators';
import isConstant from './isConstant';
import isCustomWidget from './isCustomWidget';
import isFixedItems from './isFixedItems';
import isObject from './isObject';
import labelValue from './labelValue';
import localToUTC from './localToUTC';
import mergeDefaultsWithFormData from './mergeDefaultsWithFormData';
import mergeObjects from './mergeObjects';
import mergeSchemas from './mergeSchemas';
import optionsList from './optionsList';
import orderProperties from './orderProperties';
import pad from './pad';
import parseDateString from './parseDateString';
import rangeSpec from './rangeSpec';
import replaceStringParameters from './replaceStringParameters';
import schemaRequiresTrueValue from './schemaRequiresTrueValue';
import shouldRender from './shouldRender';
import toConstant from './toConstant';
import toDateString from './toDateString';
import toErrorList from './toErrorList';
import toErrorSchema from './toErrorSchema';
import unwrapErrorHandler from './unwrapErrorHandler';
import utcToLocal from './utcToLocal';
import validationDataMerge from './validationDataMerge';
import withIdRefPrefix from './withIdRefPrefix';
import getOptionMatchingSimpleDiscriminator from './getOptionMatchingSimpleDiscriminator';

export * from './types';
export * from './enums';

export * from './constants';
export * from './parser';
export * from './schema';

export {
  allowAdditionalItems,
  ariaDescribedByIds,
  asNumber,
  canExpand,
  createErrorHandler,
  createSchemaUtils,
  dataURItoBlob,
  deepEquals,
  descriptionId,
  englishStringTranslator,
  enumOptionsDeselectValue,
  enumOptionsIndexForValue,
  enumOptionsIsSelected,
  enumOptionsSelectValue,
  enumOptionsValueForIndex,
  errorId,
  examplesId,
  ErrorSchemaBuilder,
  findSchemaDefinition,
  getDiscriminatorFieldFromSchema,
  getInputProps,
  getOptionMatchingSimpleDiscriminator,
  getSchemaType,
  getSubmitButtonOptions,
  getTemplate,
  getUiOptions,
  getWidget,
  guessType,
  hasWidget,
  hashForSchema,
  helpId,
  isConstant,
  isCustomWidget,
  isFixedItems,
  isObject,
  labelValue,
  localToUTC,
  mergeDefaultsWithFormData,
  mergeObjects,
  mergeSchemas,
  optionId,
  optionsList,
  orderProperties,
  pad,
  parseDateString,
  rangeSpec,
  replaceStringParameters,
  schemaRequiresTrueValue,
  shouldRender,
  titleId,
  toConstant,
  toDateString,
  toErrorList,
  toErrorSchema,
  unwrapErrorHandler,
  utcToLocal,
  validationDataMerge,
  withIdRefPrefix,
};
