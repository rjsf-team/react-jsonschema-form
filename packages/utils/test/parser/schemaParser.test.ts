import { schemaParser } from '../../src';
import {
  PROPERTY_DEPENDENCIES,
  RECURSIVE_REF,
  RECURSIVE_REF_ALLOF,
  SCHEMA_DEPENDENCIES,
  SCHEMA_WITH_ALLOF_CANNOT_MERGE,
  SCHEMA_AND_ONEOF_REF_DEPENDENCIES,
  SCHEMA_AND_REQUIRED_DEPENDENCIES,
  SCHEMA_WITH_ARRAY_CONDITION,
  SCHEMA_WITH_ONEOF_NESTED_DEPENDENCIES,
  SCHEMA_WITH_SINGLE_CONDITION,
  SCHEMA_WITH_MULTIPLE_CONDITIONS,
  SCHEMA_WITH_NESTED_CONDITIONS,
  SUPER_SCHEMA,
} from '../testUtils/testData';

describe('schemaParser()', () => {
  it('parses property dependencies properly', () => {
    const schemaMap = schemaParser(PROPERTY_DEPENDENCIES);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses recursive refs properly', () => {
    const schemaMap = schemaParser(RECURSIVE_REF);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses recursive refs in allOf properly', () => {
    const schemaMap = schemaParser(RECURSIVE_REF_ALLOF);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema dependencies properly', () => {
    const schemaMap = schemaParser(SCHEMA_DEPENDENCIES);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema dependencies with one of and refs properly', () => {
    const schemaMap = schemaParser(SCHEMA_AND_ONEOF_REF_DEPENDENCIES);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema and required dependencies properly', () => {
    const schemaMap = schemaParser(SCHEMA_AND_REQUIRED_DEPENDENCIES);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema with oneof and nested dependencies', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_ONEOF_NESTED_DEPENDENCIES);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema with a single condition', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_SINGLE_CONDITION);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema with multiple conditions', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_MULTIPLE_CONDITIONS);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses schema with nested conditions', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_NESTED_CONDITIONS);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parses superSchema properly', () => {
    const schemaMap = schemaParser(SUPER_SCHEMA);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parse schema with array condition', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_ARRAY_CONDITION);
    expect(schemaMap).toMatchSnapshot();
  });
  it('parse schema with allof not able to merge', () => {
    const schemaMap = schemaParser(SCHEMA_WITH_ALLOF_CANNOT_MERGE);
    expect(schemaMap).toMatchSnapshot();
  });
});
