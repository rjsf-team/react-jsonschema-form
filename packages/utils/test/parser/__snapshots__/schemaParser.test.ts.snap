// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaParser() parse schema with allof not able to merge 1`] = `
Object {
  "-44c06f09": Object {
    "$id": "-44c06f09",
    "allOf": Array [
      Object {
        "if": Object {
          "properties": Object {
            "animal": Object {
              "const": "Cat",
            },
          },
        },
        "then": Object {
          "properties": Object {
            "food": Object {
              "enum": Array [
                "meat",
                "grass",
                "fish",
              ],
              "type": "string",
            },
          },
          "required": Array [
            "food",
          ],
        },
      },
      Object {
        "if": Object {
          "properties": Object {
            "animal": Object {
              "const": "Fish",
            },
          },
        },
        "then": Object {
          "properties": Object {
            "food": Object {
              "enum": Array [
                "insect",
                "worms",
              ],
              "type": "string",
            },
            "water": Object {
              "enum": Array [
                "lake",
                "sea",
              ],
              "type": "string",
            },
          },
          "required": Array [
            "food",
            "water",
          ],
        },
      },
      Object {
        "required": Array [
          "animal",
        ],
      },
    ],
    "properties": Object {
      "animal": Object {
        "enum": Array [
          "Cat",
          "Fish",
        ],
      },
    },
    "type": "object",
  },
  "59fefb6e": Object {
    "$id": "59fefb6e",
    "properties": Object {
      "animal": Object {
        "const": "Fish",
      },
    },
  },
  "6f438756": Object {
    "$id": "6f438756",
    "properties": Object {
      "animal": Object {
        "const": "Cat",
      },
    },
  },
}
`;

exports[`schemaParser() parse schema with array condition 1`] = `
Object {
  "-33d8d73c": Object {
    "$id": "-33d8d73c",
    "properties": Object {
      "country": Object {
        "const": "United States of America",
      },
    },
  },
  "f254d0a": Object {
    "$id": "f254d0a",
    "properties": Object {
      "list": Object {
        "items": Object {
          "else": Object {
            "properties": Object {
              "postal_code": Object {
                "pattern": "[A-Z][0-9][A-Z] [0-9][A-Z][0-9]",
              },
            },
          },
          "if": Object {
            "properties": Object {
              "country": Object {
                "const": "United States of America",
              },
            },
          },
          "properties": Object {
            "country": Object {
              "default": "United States of America",
              "enum": Array [
                "United States of America",
                "Canada",
              ],
            },
          },
          "then": Object {
            "properties": Object {
              "postal_code": Object {
                "pattern": "[0-9]{5}(-[0-9]{4})?",
              },
            },
          },
          "type": "object",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses property dependencies properly 1`] = `
Object {
  "-4e612676": Object {
    "$id": "-4e612676",
    "dependencies": Object {
      "a": Array [
        "b",
      ],
    },
    "properties": Object {
      "a": Object {
        "type": "string",
      },
      "b": Object {
        "type": "integer",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
}
`;

exports[`schemaParser() parses recursive refs in allOf properly 1`] = `
Object {
  "2b3d97f8": Object {
    "$id": "2b3d97f8",
    "definitions": Object {
      "@enum": Object {
        "properties": Object {
          "_id": Object {
            "title": "Value",
            "type": "number",
          },
          "children": Object {
            "items": Object {
              "allOf": Array [
                Object {
                  "$ref": "#/definitions/@enum",
                },
              ],
            },
            "title": "Subvalues",
            "type": "array",
          },
          "name": Object {
            "default": "",
            "title": "Name",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "properties": Object {
      "value": Object {
        "items": Object {
          "allOf": Array [
            Object {
              "$ref": "#/definitions/@enum",
            },
          ],
        },
        "minItems": 1,
        "type": "array",
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses recursive refs properly 1`] = `
Object {
  "7cf4cd8a": Object {
    "$id": "7cf4cd8a",
    "$ref": "#/definitions/@enum",
    "definitions": Object {
      "@enum": Object {
        "properties": Object {
          "children": Object {
            "$ref": "#/definitions/@enum",
          },
          "name": Object {
            "default": "",
            "title": "Name",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
}
`;

exports[`schemaParser() parses schema and required dependencies properly 1`] = `
Object {
  "-738b34fc": Object {
    "$id": "-738b34fc",
    "dependencies": Object {
      "a": Object {
        "properties": Object {
          "a": Object {
            "type": "string",
          },
        },
        "required": Array [
          "b",
        ],
      },
    },
    "properties": Object {
      "a": Object {
        "type": "string",
      },
      "b": Object {
        "type": "integer",
      },
    },
    "required": Array [
      "a",
    ],
    "type": "object",
  },
}
`;

exports[`schemaParser() parses schema dependencies properly 1`] = `
Object {
  "75f1f611": Object {
    "$id": "75f1f611",
    "dependencies": Object {
      "a": Object {
        "properties": Object {
          "b": Object {
            "type": "integer",
          },
        },
      },
    },
    "properties": Object {
      "a": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses schema dependencies with one of and refs properly 1`] = `
Object {
  "-29b1244f": Object {
    "$id": "-29b1244f",
    "properties": Object {
      "a": Object {
        "enum": Array [
          "int",
        ],
      },
    },
    "type": "object",
  },
  "71dd1668": Object {
    "$id": "71dd1668",
    "properties": Object {
      "a": Object {
        "enum": Array [
          "bool",
        ],
      },
    },
    "type": "object",
  },
  "7de97d9a": Object {
    "$id": "7de97d9a",
    "definitions": Object {
      "needsA": Object {
        "properties": Object {
          "a": Object {
            "enum": Array [
              "int",
            ],
          },
          "b": Object {
            "type": "integer",
          },
        },
      },
      "needsB": Object {
        "properties": Object {
          "a": Object {
            "enum": Array [
              "bool",
            ],
          },
          "b": Object {
            "type": "boolean",
          },
        },
      },
    },
    "dependencies": Object {
      "a": Object {
        "oneOf": Array [
          Object {
            "$ref": "#/definitions/needsA",
          },
          Object {
            "$ref": "#/definitions/needsB",
          },
        ],
      },
    },
    "properties": Object {
      "a": Object {
        "enum": Array [
          "int",
          "bool",
        ],
        "type": "string",
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses schema with a single condition 1`] = `
Object {
  "-33d8d73c": Object {
    "$id": "-33d8d73c",
    "properties": Object {
      "country": Object {
        "const": "United States of America",
      },
    },
  },
  "7a88ed97": Object {
    "$id": "7a88ed97",
    "else": Object {
      "properties": Object {
        "postal_code": Object {
          "pattern": "[A-Z][0-9][A-Z] [0-9][A-Z][0-9]",
        },
      },
    },
    "if": Object {
      "properties": Object {
        "country": Object {
          "const": "United States of America",
        },
      },
    },
    "properties": Object {
      "country": Object {
        "default": "United States of America",
        "enum": Array [
          "United States of America",
          "Canada",
        ],
      },
    },
    "then": Object {
      "properties": Object {
        "postal_code": Object {
          "pattern": "[0-9]{5}(-[0-9]{4})?",
        },
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses schema with multiple conditions 1`] = `
Object {
  "-269ce7ef": Object {
    "$id": "-269ce7ef",
    "properties": Object {
      "Animal": Object {
        "const": "Dog",
      },
    },
    "required": Array [
      "Animal",
    ],
  },
  "-5e82ba72": Object {
    "$id": "-5e82ba72",
    "properties": Object {
      "BreedName": Object {
        "const": "Dalmation",
      },
    },
    "required": Array [
      "BreedName",
    ],
  },
  "-61d82723": Object {
    "$id": "-61d82723",
    "allOf": Array [
      Object {
        "if": Object {
          "properties": Object {
            "Animal": Object {
              "const": "Cat",
            },
          },
          "required": Array [
            "Animal",
          ],
        },
        "then": Object {
          "properties": Object {
            "Tail": Object {
              "default": "Long",
              "enum": Array [
                "Long",
                "Short",
                "None",
              ],
              "title": "Tail length",
              "type": "string",
            },
          },
          "required": Array [
            "Tail",
          ],
        },
      },
      Object {
        "if": Object {
          "properties": Object {
            "Animal": Object {
              "const": "Dog",
            },
          },
          "required": Array [
            "Animal",
          ],
        },
        "then": Object {
          "properties": Object {
            "Breed": Object {
              "allOf": Array [
                Object {
                  "if": Object {
                    "properties": Object {
                      "BreedName": Object {
                        "const": "Alsatian",
                      },
                    },
                    "required": Array [
                      "BreedName",
                    ],
                  },
                  "then": Object {
                    "properties": Object {
                      "Fur": Object {
                        "default": "brown",
                        "enum": Array [
                          "black",
                          "brown",
                        ],
                        "title": "Fur",
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "Fur",
                    ],
                  },
                },
                Object {
                  "if": Object {
                    "properties": Object {
                      "BreedName": Object {
                        "const": "Dalmation",
                      },
                    },
                    "required": Array [
                      "BreedName",
                    ],
                  },
                  "then": Object {
                    "properties": Object {
                      "Spots": Object {
                        "default": "small",
                        "enum": Array [
                          "large",
                          "small",
                        ],
                        "title": "Spots",
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "Spots",
                    ],
                  },
                },
              ],
              "properties": Object {
                "BreedName": Object {
                  "default": "Alsatian",
                  "enum": Array [
                    "Alsatian",
                    "Dalmation",
                  ],
                  "title": "Breed name",
                  "type": "string",
                },
              },
              "required": Array [
                "BreedName",
              ],
              "title": "Breed",
            },
          },
        },
      },
    ],
    "properties": Object {
      "Animal": Object {
        "default": "Cat",
        "enum": Array [
          "Cat",
          "Dog",
        ],
        "title": "Animal",
        "type": "string",
      },
    },
    "required": Array [
      "Animal",
    ],
    "type": "object",
  },
  "-7b9d5dae": Object {
    "$id": "-7b9d5dae",
    "properties": Object {
      "BreedName": Object {
        "const": "Alsatian",
      },
    },
    "required": Array [
      "BreedName",
    ],
  },
  "6dc2a62b": Object {
    "$id": "6dc2a62b",
    "properties": Object {
      "Animal": Object {
        "const": "Cat",
      },
    },
    "required": Array [
      "Animal",
    ],
  },
}
`;

exports[`schemaParser() parses schema with nested conditions 1`] = `
Object {
  "-f42a226": Object {
    "$id": "-f42a226",
    "properties": Object {
      "state": Object {
        "const": "New York",
      },
    },
    "required": Array [
      "state",
    ],
  },
  "14f43ce2": Object {
    "$id": "14f43ce2",
    "properties": Object {
      "country": Object {
        "const": "USA",
      },
    },
    "required": Array [
      "country",
    ],
  },
  "5ad8e931": Object {
    "$id": "5ad8e931",
    "properties": Object {
      "state": Object {
        "const": "California",
      },
    },
    "required": Array [
      "state",
    ],
  },
  "767ff7c6": Object {
    "$id": "767ff7c6",
    "if": Object {
      "properties": Object {
        "country": Object {
          "const": "USA",
        },
      },
      "required": Array [
        "country",
      ],
    },
    "properties": Object {
      "country": Object {
        "enum": Array [
          "USA",
        ],
      },
    },
    "required": Array [
      "country",
    ],
    "then": Object {
      "else": Object {
        "if": Object {
          "properties": Object {
            "state": Object {
              "const": "California",
            },
          },
          "required": Array [
            "state",
          ],
        },
        "then": Object {
          "properties": Object {
            "city": Object {
              "enum": Array [
                "Los Angeles",
                "San Diego",
                "San Jose",
              ],
              "type": "string",
            },
          },
        },
      },
      "if": Object {
        "properties": Object {
          "state": Object {
            "const": "New York",
          },
        },
        "required": Array [
          "state",
        ],
      },
      "properties": Object {
        "state": Object {
          "enum": Array [
            "California",
            "New York",
          ],
          "type": "string",
        },
      },
      "required": Array [
        "state",
      ],
      "then": Object {
        "properties": Object {
          "city": Object {
            "enum": Array [
              "New York City",
              "Buffalo",
              "Rochester",
            ],
            "type": "string",
          },
        },
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses schema with oneof and nested dependencies 1`] = `
Object {
  "-37c29d49": Object {
    "$id": "-37c29d49",
    "properties": Object {
      "update_absences": Object {
        "const": "NON_MEDICAL_ONLY",
      },
    },
    "type": "object",
  },
  "-406a5a54": Object {
    "$id": "-406a5a54",
    "dependencies": Object {
      "employee_accounts": Object {
        "oneOf": Array [
          Object {
            "properties": Object {
              "employee_accounts": Object {
                "const": true,
              },
              "update_absences": Object {
                "oneOf": Array [
                  Object {
                    "const": "BOTH",
                    "title": "Both",
                  },
                ],
                "title": "Update Absences",
                "type": "string",
              },
            },
          },
        ],
      },
      "update_absences": Object {
        "oneOf": Array [
          Object {
            "properties": Object {
              "permitted_extension": Object {
                "title": "Permitted Extension",
                "type": "integer",
              },
              "update_absences": Object {
                "const": "BOTH",
              },
            },
          },
          Object {
            "properties": Object {
              "permitted_extension": Object {
                "title": "Permitted Extension",
                "type": "integer",
              },
              "update_absences": Object {
                "const": "MEDICAL_ONLY",
              },
            },
          },
          Object {
            "properties": Object {
              "permitted_extension": Object {
                "title": "Permitted Extension",
                "type": "integer",
              },
              "update_absences": Object {
                "const": "NON_MEDICAL_ONLY",
              },
            },
          },
        ],
      },
    },
    "properties": Object {
      "employee_accounts": Object {
        "title": "Employee Accounts",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "-7a6dd81b": Object {
    "$id": "-7a6dd81b",
    "properties": Object {
      "update_absences": Object {
        "const": "MEDICAL_ONLY",
      },
    },
    "type": "object",
  },
  "-7c3c9229": Object {
    "$id": "-7c3c9229",
    "const": "BOTH",
    "title": "Both",
  },
  "3a28e528": Object {
    "$id": "3a28e528",
    "properties": Object {
      "employee_accounts": Object {
        "const": true,
      },
    },
    "type": "object",
  },
  "934b22c": Object {
    "$id": "934b22c",
    "properties": Object {
      "update_absences": Object {
        "const": "BOTH",
      },
    },
    "type": "object",
  },
}
`;

exports[`schemaParser() parses superSchema properly 1`] = `
Object {
  "-130bae51": Object {
    "$id": "-130bae51",
    "anyOf": Array [
      Object {
        "required": Array [
          "name",
        ],
      },
    ],
    "properties": Object {
      "name": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "-1e3c57ec": Object {
    "$id": "-1e3c57ec",
    "anyOf": Array [
      Object {
        "required": Array [
          "choice",
        ],
      },
      Object {
        "required": Array [
          "more",
        ],
      },
    ],
    "properties": Object {
      "choice": Object {
        "const": "two",
        "type": "string",
      },
      "more": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "-42566fe2": Object {
    "$id": "-42566fe2",
    "anyOf": Array [
      Object {
        "required": Array [
          "choice",
        ],
      },
      Object {
        "required": Array [
          "other",
        ],
      },
    ],
    "properties": Object {
      "choice": Object {
        "const": "one",
        "type": "string",
      },
      "other": Object {
        "type": "number",
      },
    },
    "type": "object",
  },
  "super-schema": Object {
    "$id": "super-schema",
    "definitions": Object {
      "choice1": Object {
        "properties": Object {
          "choice": Object {
            "const": "one",
            "type": "string",
          },
          "other": Object {
            "type": "number",
          },
        },
        "type": "object",
      },
      "choice2": Object {
        "properties": Object {
          "choice": Object {
            "const": "two",
            "type": "string",
          },
          "more": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "foo": Object {
        "properties": Object {
          "name": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "list": Object {
        "items": Object {
          "type": "string",
        },
        "type": "array",
      },
      "passwords": Object {
        "properties": Object {
          "pass1": Object {
            "type": "string",
          },
          "pass2": Object {
            "type": "string",
          },
        },
        "required": Array [
          "pass1",
          "pass2",
        ],
        "type": "object",
      },
      "price": Object {
        "minimum": 0,
        "multipleOf": 0.03,
        "title": "Price per task ($)",
        "type": "number",
      },
    },
    "properties": Object {
      "anything": Object {
        "additionalProperties": Object {
          "type": "string",
        },
        "type": "object",
      },
      "dataUrlWithName": Object {
        "format": "data-url",
        "type": "string",
      },
      "foo": Object {
        "type": "string",
      },
      "list": Object {
        "$ref": "#/definitions/list",
      },
      "multi": Object {
        "anyOf": Array [
          Object {
            "$ref": "#/definitions/foo",
          },
        ],
      },
      "passwords": Object {
        "$ref": "#/definitions/passwords",
      },
      "price": Object {
        "$ref": "#/definitions/price",
      },
      "single": Object {
        "oneOf": Array [
          Object {
            "$ref": "#/definitions/choice1",
          },
          Object {
            "$ref": "#/definitions/choice2",
          },
        ],
      },
    },
    "type": "object",
  },
}
`;
